tests/unit/s2n_self_talk_client_hello_cb_test.c

```diff
@@ -0,0 +1,319 @@
+    "ou1fVWXOJ1kealYPbRPj8pDa31omB8q5Cf8Qe/b9anqyi9CsP17QbVg9k2IgoLlj\n"
+    "agqOc0u/opOTZB4tqJbqsIdEhc5LD5RUkYJsw00Iq0RSiKTfiWSPyOFw99Y9Act0\n"
+    "cbIIxEECgYEA8/SOsQjoUX1ipRvPbfO3suV1tU1hLCQbIpv7WpjNr1kHtngjzQMP\n"
+    "dU/iriUPGF1H+AxJJcJQfCVThV1AwFYVKb/LCrjaxlneZSbwfehpjo+xQGaNYG7Q\n"
+    "1vQuBVejuYk/IvpZltQOdm838DjvYyWDMh4dcMFIycXxEg+oHxf/s+8CgYEA0n4p\n"
+    "GBuLUNx9vv3e84BcarLaOF7wY7tb8z2oC/mXztMZpKjovTH0PvePgI5/b3KQ52R0\n"
+    "8zXHVX/4lSQVtCuhOVwKOCQq97/Zhlp5oTTShdQ0Qa1GQRl5wbTS6hrYEWSi9AQP\n"
+    "BVUPZ+RIcxx00DfBNURkId8xEpvCOmvySN8sUlUCgYAtXmHbEqkB3qulwRJGhHi5\n"
+    "UGsfmJBlwSE6wn9wTdKStZ/1k0o1KkiJrJ2ffUzdXxuvSbmgyA5nyBlMSBdurZOp\n"
+    "+/0qtU4abUQq058OC1b2KEryix/nuzQjha25WJ8eNiQDwUNABZfa9rwUdMIwUh2g\n"
+    "CHG5Mnjy7Vjz3u2JOtFXCQKBgQCVRo1EIHyLauLuaMINM9HWhWJGqeWXBM8v0GD1\n"
+    "pRsovQKpiHQNgHizkwM861GqqrfisZZSyKfFlcynkACoVmyu7fv9VoD2VCMiqdUq\n"
+    "IvjNmfE5RnXVQwja+668AS+MHi+GF77DTFBxoC5VHDAnXfLyIL9WWh9GEBoNLnKT\n"
+    "hVm8RQKBgQCB9Skzdftc+14a4Vj3NCgdHZHz9mcdPhzJXUiQyZ3tYhaytX9E8mWq\n"
+    "pm/OFqahbxw6EQd86mgANBMKayD6B1Id1INqtXN1XYI50bSs1D2nOGsBM7MK9aWD\n"
+    "JXlJ2hwsIc4q9En/LR3GtBaL84xTHGfznNylNhXi7GbO1wNMJuAukA==\n"
+    "-----END RSA PRIVATE KEY-----\n";
+
+int mock_client(int writefd, int readfd, int expect_failure)
+{
+    struct s2n_connection *conn;
+    s2n_blocked_status blocked;
+    int result = 0;
+    int rc = 0;
+
+    /* Give the server a chance to listen */
+    sleep(1);
+
+    conn = s2n_connection_new(S2N_CLIENT);
+    conn->server_protocol_version = S2N_TLS12;
+    conn->client_protocol_version = S2N_TLS12;
+    conn->actual_protocol_version = S2N_TLS12;
+
+    s2n_connection_set_read_fd(conn, readfd);
+    s2n_connection_set_write_fd(conn, writefd);
+
+    s2n_set_server_name(conn, "example.com");
+
+    rc = s2n_negotiate(conn, &blocked);
+    if (expect_failure) {
+        if (!rc) {
+            result = 1;
+        }
+
+        if (s2n_connection_get_alert(conn) != 40){
+            result = 2;
+        }
+    } else {
+        char buffer[0xffff];
+        if (rc < 0) {
+            result = 1;
+        }
+
+        for (int i = 1; i < 0xffff; i += 100) {
+            memset(buffer, 33, sizeof(char) * i);
+            s2n_send(conn, buffer, i, &blocked);
+        }
+
+        int shutdown_rc= -1;
+        do {
+            shutdown_rc = s2n_shutdown(conn, &blocked);
+        } while(shutdown_rc != 0);
+    }
+
+    s2n_connection_free(conn);
+
+    /* Give the server a chance to a void a sigpipe */
+    sleep(1);
+
+    _exit(result);
+}
+
+int mock_nanoseconds_since_epoch(void *data, uint64_t *nanoseconds)
+{
+    static int called = 0;
+
+    /* When first called return 0 seconds */
+    *nanoseconds = 0;
+
+    /* When next called return 31 seconds */
+    if (called) {
+        *nanoseconds += (uint64_t) 31 * 1000000000;
+    }
+
+    called = 1;
+
+    return 0;
+}
+
+int client_hello_swap_config(struct s2n_connection *conn, void *ctx)
```

Comment:
how about you increment a counter or something in here so you can validate it was called in your test?

perhaps wrap your swap_config in a struct of

{
  int was_invoked;
  s2n_config *config;
}


then in here you could set was_invoked to 1, and validate that in the test after handhake is done that we actually called this?